staging/src/k8s.io/apimachinery/pkg/api/equality/semantic.go

```diff
@@ -46,4 +49,14 @@ var Semantic = conversion.EqualitiesOrDie(
 	func(a, b fields.Selector) bool {
 		return a.String() == b.String()
 	},
+	func(a, b runtime.RawExtension) bool {
+		rawa, rawb := a.Raw, b.Raw
+		if jsona, err := a.MarshalJSON(); err == nil {
+			rawa = jsona
+		}
+		if jsonb, err := b.MarshalJSON(); err == nil {
+			rawb = jsonb
+		}
+		return bytes.Equal(rawa, rawb)
+	},
```

Comment:
This is the scenario:

1. Server accepts either JSON or CBOR request encoding.
2. A client (dynamic or typed) creates an object using JSON as its request encoding. The request bytes that represent the RawExtension field must be JSON-encoded.
3. The server decodes into the RawExtension field, its `Raw` bytes are JSON, and the bytes are persisted.
4. The same client from (2) changes its configuration so that it prefers CBOR request encoding. It makes an update request to the resource it created.
5. This time, when the server decodes into RawExtension, the `Raw` bytes must be CBOR.
6. Even though the Go object that the client encoded to the request body did not change, the change in its wire representation changes the bytes of the RawExtension that the server sees.

Is it up to each apiserver to handle this on a case-by-case basis when serving a resource that uses RawExtension?

Reply:
What are the constraints here? By RawExtension go doc, every RawExtension is paired with a `runtime.Object` in the internal types. I.e. the server would not roundtrip directly through JSON/CBOR only, but through an intermediate native type. Hence, for what is stored in etcd the content type has no influence.

Reply:
At least that's the theory. Which other cases do we have? Is the event one like that?

Reply:
ControllerRevision in apps/v1 currently represents the same field as RawExtension in its external and internal versions, but also, the default RawExtension->Object conversion produces an Unknown that preserves the exact serialization from the RawExtension.

Is this sort of issue ultimately arising from conversions whose result depends on insignificant serialization differences (e.g. for JSON, whitespace or object key order)?

Reply:
> By RawExtension go doc, every RawExtension is paired with a runtime.Object in the internal types. I.e. the server would NOT roundtrip directly through JSON/CBOR only, but through an intermediate native type.

In practice, we only had a single non-experimental RawExtension use in in-tree types (ControllerRevision), which round-tripped through runtime.Unknown which preserves byte-for-byte.

That actually caused multiple *problems*:
* Internal round-tripping does not normalize formatting, but user-round-tripping through JSON can, and update validation requires the bytes in the field to be immutable. That means if one user created the object, another user read/wrote JSON and normalized the order, it could fail update validation because the format of the field changed.
* Prior to https://github.com/kubernetes/kubernetes/pull/125549, nothing ensured the content in the field (when written via a proto request) was even valid JSON. The byte-preserving round-tripping through runtime.Object / runtime.Unknown didn't do anything to normalize or validate or convert.

There was essentially no benefit to the opaque internal runtime.Object field, since it prevented inspection and validation of the bytes.